### PR TITLE
Improve foreign key introspection on SQLite

### PR DESCRIPTION
**Pull request by morozov**
|      Q       |   A
|------------- | -----------
| Type         | improvement, deprecation

With this patch, DBAL can support all foreign key operations except for altering foreign keys on an existing table. IMO, it's still sufficient to declare that the platform supports foreign keys.

This might be considered a minor breaking change (see [xkcd #1172](https://xkcd.com/1172/)) in the cases where code that alters a foreign key based on whether the current platform supports foreign keys. I'd argue that it's a made up/synthetic scenario that primarily applies to testing the DBAL or other libraries that provide an abstraction on top of the DBAL (e.g. the ORM).

In real practical cases, the application would use a subset of the features supported by all target platforms. So it would either not attempt to alter foreign keys because SQLite doesn't support that or it wouldn't support SQLite. In such cases, this change is backward-compatible.

UPD: it still looks like it will cause the same failure in the Laravel test suite that https://github.com/doctrine/dbal/pull/3762 caused since the test suite relies on `AbstractPlatform::supportsForeignKeyConstraints()` (see https://github.com/laravel/framework/issues/34093#issuecomment-685719863).

### CODE REVIEW

**Comment by greg0ire** [comment]
Should https://github.com/doctrine/dbal/pull/4258 be cherry-picked into this?

**Comment by morozov** [comment]
> Should #4258 be cherry-picked into this?

It looks like it depends on `supportsCreateDropForeignKeyConstraints()`. I don't believe we should introduce any new `supports*()` method. This is exactly the reason why such changes may be considered BC breaks.

**Comment by greg0ire** [comment]
> the test suite relies on AbstractPlatform::supportsForeignKeyConstraints()

Maybe that's indirect somehow because `git log -S supportsForeignKeyConstraints` from the git repo of Laravel yields nothing…

@driesvints hi! :wave: Can you please review this, and maybe try it on the Laravel test suite? So that we can know if something is going to break and give you some time to prevent it?

File: src/Schema/AbstractSchemaManager.php [hunk 1]
```diff
@@ -521,14 +507,56 @@ protected function selectForeignKeyColumns(string $databaseName, ?string $tableN
...
     /**
-     * Returns table options for the tables in the specified database. If the table name is specified, narrows down
-     * the selection to this table.
+     * Fetches definitions of table columns in the specified database and returns them grouped by table name.
+     *
+     * @return array<string,list<array<string,mixed>>>
+     *
+     * @throws Exception
+     */
+    protected function fetchTableColumnsByTable(string $databaseName): array
```
Comment (greg0ire) [inline src/Schema/AbstractSchemaManager.php:516]: Should some of these new methods be `final`? They don't look like they are intended to be overridden.

Comment (morozov) [reply src/Schema/AbstractSchemaManager.php:516]: The point of having these methods extracted (specifically, `fetchForeignKeyColumnsByTable()`) is being able to do an override for SQLite. As the next step, I want to introduce a separate API for schema introspection, so we're likely to deprecate these methods in the future.

**Comment by morozov** [comment]
> Maybe that's indirect somehow because `git log -S supportsForeignKeyConstraints` from the git repo of Laravel yields nothing…

I did the same and also didn't find anything. Having somebody from Laravel take a look would be nice.

**Comment by driesvints** [comment]
Hi all. I tried these changes out on Laravel 9.x's test suite and everything still passed so I think we're good. Of course there could always be edge cases in apps.

**Review by greg0ire — APPROVED** [review]

**Comment by greg0ire** [comment]
Great! Thanks for taking the time to check!

**Review by derrabus — APPROVED** [review]

**Comment by greg0ire** [comment]
@morozov @derrabus I think this breaks the ORM test suite: https://github.com/doctrine/orm/runs/6563292058?check_suite_focus=true  :disappointed:

**Comment by morozov** [comment]
It looks like the ORM always creates foreign keys via the ALTER statements as opposed to creating the table and the foreign keys at once. As a temporary fix, we could replace `if supports foreign keys` with `if supports foreign keys and not sqlite`. But it should be possible to fix the ORM property.

**Comment by morozov** [comment]
I believe I see what's going on here. When it comes to creating a schema where some tables contain foreign keys, neither ORM nor DBAL know in which order they need to be created in order to have the foreign key dependencies satisfied at every step of the migration, so the current implementation does this:

1. There's a flag in `AbstractPlatform::getCreateTableSQL()`: https://github.com/doctrine/dbal/blob/9b5215a27a76f2f540906e7d69548e82f0af74f7/src/Platforms/AbstractPlatform.php#L1928 The "indexes" part here is for show, nobody wants to disable creating indexes. The signature should be read as "do not create foreign keys by default". Note that SQLite cannot create foreign keys afterwards and it doesn't support them at all, so `$createFlags` for SQLite include foreign keys as well: https://github.com/doctrine/dbal/blob/7df5dd5cb871d8451ef8c648ad6bb518cba51bd8/src/Platforms/SqlitePlatform.php#L884-L886
3. When accepting a visitor, the `Table` class not only passes itself to the visitor but also passes its columns, indices and foreign keys: https://github.com/doctrine/dbal/blob/83d1f7032375c995c214426896e17069bd59ad69/src/Schema/Table.php#L911-L926 Why? Doesn't the table itself contain them? You should already know the answer. Because we want to create tables and foreign keys separately in `CreateSchemaSqlCollector` which conveniently doesn't implement `acceptColumn()` or `acceptIndex()` but does implement `acceptForeignKey()`.
4. Note that the Schema Manager API doesn't support creating schemas (the `createSchema()` method doesn't do what you think it does). It's the `CreateSchemaSqlCollector` visitor which attempts to accept a schema and generate SQL for it.

In order to unscrew this all we need to:

1. Have the collector create tables in the right order in order to be able to create foreign keys right away.
2. Have the collector create tables together with the foreign keys. This would be a breaking change 💣 because then it will create the keys again and fail because they already exist.
3. Have `Table` not pass its columns, indexes and foreign keys to the visitor explicitly and expect visitors to process them all processing the table itself. This would be a breaking change 💣 again.
4. Remove `acceptColumn()`, etc from the Visitor interface because they are part of the table. What the heck does accepting a column mean if we don't even know which table it belongs to? Again, breaking change 💣.
5. Drop the support for `$createFlags` because it should be no longer necessary. Yes, this is also a breaking change 💣.

Alternatively, we could move the logic of creating schemas to the schema manager where it belongs (how would we name the method?). Then we could ditch all this visitor thing in one shot.

**Comment by greg0ire** [comment]
Thanks for the in-depth analysis @morozov! 

Deprecating the collector in favor of a new schema manager method sounds more sensible

> how would we name the method?

`deploySchema`?
